### ISSUE TITLE
[DataGrid] Introduces `minWidth` in `ColDef` for columns with fluid width 

### ISSUE THREAD

**Issue by DanailH** [mui-org/material-ui-x#1850]
Based on these discussions:
https://github.com/mui-org/material-ui-x/issues/1561#issuecomment-832753596
https://github.com/mui-org/material-ui-x/pull/1816#issuecomment-855771150

## Summary 💡

Have the ability to set `minWidth` for columns.

## Examples 🌈

A possible solution can be the following (adding optional `minWidth` for fluid columns):

```tsx
<DataGrid
  columns={[
    {
      field: 'id',
      flex: 1,
      minWidth: 50,
    },
    {
      field: 'username',
      width: 200,
    },
    {
      field: 'age',
      flex: 0.3,
      minWidth: 80,
    },
  ]}
  rows={{
    id: 1,
    username: '@MaterialUI',
    age: 20,
  }}
 />
```

## Motivation 🔦

In order for the fluid column width feature to be consistent, we need to introduce a way to set the `minWidth` of a fluid column.

## Benchmark

- https://ag-grid.com/javascript-grid/column-sizing/#column-flex

**Comment by oliviertassinari** [comment]
We will need to revert https://github.com/mui-org/material-ui-x/pull/1816 committed in HEAD under ab0f41222aa0caf3ff7f2bfbbfa369a5fdca48f4 as per point 2 https://github.com/mui-org/material-ui-x/pull/1816#issuecomment-855771150. It breaks the consistency of the behavior of the columns

**Comment by DanailH** [comment]
@mui-org/x  I need quick feedback - if the `minWidth` is provided should it be respected when resizing the column?

I'm asking because currently there is a strict minimum width of `50px` when resizing a column.

**Comment by flaviendelangle** [comment]
I think it should respect it.

**Comment by DanailH** [comment]
@flaviendelangle I think the same but still, there are drawbacks - if column width is `100px` for example and you set the minWidth to `100px` that essentially means that you can't shrink the column, only expand it.

**Comment by oliviertassinari** [comment]
>  if the minWidth is provided should it be respected when resizing the column?

I would vote for leaving the end-users to use the resize feature freely, so allowing end-users to ignore the min-width values when resizing. You can try out the behavior when resizing is blocked by min-width https://ag-grid.com/javascript-grid/column-sizing/#column-flex, it's frustrating, and feels out of place. If we come back to why we introduce this feature, it's to fix the auto layout. It wasn't to fix an issue with resizing.

**Comment by oliviertassinari** [comment]
Staying on the matter of resizing + flex, we lost the resized state when the size of the container changes. See


https://user-images.githubusercontent.com/3165635/125062727-01430480-e0af-11eb-871f-131328f4e3f6.mp4

https://codesandbox.io/s/material-demo-forked-ch5hy?file=/demo.tsx

vs.

https://user-images.githubusercontent.com/3165635/125063170-70b8f400-e0af-11eb-857f-7482d4f28a6f.mp4

https://plnkr.co/edit/VO2y32CCitTxbgar

Should I open a new issue?
